broadinstitute/rawls
# [early] GAWB-3625 Validate MCs properly

[still todo]:
- [x] swagger
- [x] push swagger through orchestration: https://github.com/broadinstitute/firecloud-orchestration/pull/654
- [x] update fc-ui to read missingInputs and extraInputs too https://github.com/broadinstitute/firecloud-ui/pull/1322
- [ ] probably update auto-tests
- [ ] inform saturn of incoming changes

See the comments on [the JIRA ticket](https://broadinstitute.atlassian.net/browse/GAWB-3625) for more details and background here.

Current behaviour:

* Calling Rawls' "validate method configuration" endpoint checks that all expressions in your MC are valid. It does this a) for all inputs b) without checking against the WDL to see if those inputs are optional.
    * This means that MCs with empty expressions for optional inputs are marked as _invalid_.
* The "validate method configuration" codepath is called when you create or update a method configuration.
* Method configurations are saved regardless of whether their expressions are valid.
* Calling Rawls' "validate submission" endpoint _does_ check against the real method.

New behaviour:

* Both the "validate method configuration" and "validate submission" codepaths pull the WDL and validate your method configuration properly, taking into account whether your inputs are optional.
* Method configurations continue to be saved regardless of whether their expressions are valid.

The effect of this is that the `ValidatedMethodConfiguration` structure looks more like the output of "validated submissions" -- i.e. it includes information of "extra" inputs (that exist in your MC but not the WDL) and "missing" inputs (that exist in the WDL but not your MC).

This is a behavioural change to the API but not a breaking one. If you're creating or updating a MC, your status codes will remain the same. The JSON returned will be different in that there are two extra fields, but this isn't a breaking change that will force API consumers to rewrite their JSON parsing. (They _will_ see different responses but that's okay.)

  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
